stock_cancel/models/stock.py

```diff
@@ -47,15 +48,27 @@ def action_revert_done(self):
             if picking.invoice_id:
                 raise exceptions.Warning(
                     _('Picking %s has invoices!') % (picking.name))
+            for move in picking.move_lines:
+                for quant in move.quant_ids:
+                    if quant.location_id == move.location_dest_id:
+                        if quant.qty != move.product_uom_qty:
+                            raise exceptions.Warning(
+                                _('Picking %s cannot be cancelled because '
+                                  'their quants have later moves associated '
+                                  'to them.')
+                                % (picking.name))
+                        quant_obj.quants_move(
+                            [(quant, quant.qty)], move, move.location_id,
+                            location_from=move.location_dest_id,
+                            lot_id=quant.lot_id.id, owner_id=quant.owner_id.id)
+                quant_obj.quants_unreserve(move)
             picking.move_lines.write({'state': 'draft'})
             picking.state = 'draft'
             if picking.invoice_state == 'invoiced' and not picking.invoice_id:
```

Comment:
You don't need to check again `not picking.invoice_id`, as this is a warning condition